server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java

```diff
@@ -0,0 +1,167 @@
+            .blockOptional()
+            .orElseThrow(() -> new ObjectNotFoundException(String.format("Could not retrieve blob metadata for %s", blobId)));
+    }
+
+    @Override
+    public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.just(bucketName)
+            .filter(defaultBucket::equals)
+            .flatMap(ignored -> Mono.from(cache.read(blobId)))
+            .switchIfEmpty(Mono.from(backend.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public Mono<BlobId> save(BucketName bucketName, byte[] bytes, StoragePolicy storagePolicy) {
+        return Mono.from(backend.save(bucketName, bytes, storagePolicy))
+            .flatMap(blobId -> {
+                if (isAbleToCache(bucketName, bytes, storagePolicy)) {
+                    return Mono.from(cache.cache(blobId, bytes)).thenReturn(blobId);
+                }
+                return Mono.just(blobId);
+            });
+    }
+
+    @Override
+    public Publisher<BlobId> save(BucketName bucketName, InputStream inputStream, StoragePolicy storagePolicy) {
+        Preconditions.checkNotNull(inputStream, "InputStream must not be null");
+
+        if (isAbleToCache(bucketName, storagePolicy)) {
+            return saveInCache(bucketName, inputStream, storagePolicy);
+        }
+
+        return backend.save(bucketName, inputStream, storagePolicy);
+    }
+
+    private Publisher<BlobId> saveInCache(BucketName bucketName, InputStream inputStream, StoragePolicy storagePolicy) {
+        PushbackInputStream pushbackInputStream = new PushbackInputStream(inputStream, sizeThresholdInBytes + 1);
+
+        return Mono.fromCallable(() -> fullyReadSmallStream(pushbackInputStream))
+            .flatMap(Mono::justOrEmpty)
+            .filter(bytes -> isAbleToCache(bucketName, bytes, storagePolicy))
+            .flatMap(bytes -> Mono.from(backend.save(bucketName, pushbackInputStream, storagePolicy))
+                .flatMap(blobId -> Mono.from(cache.cache(blobId, bytes))
+                    .thenReturn(blobId)))
+            .switchIfEmpty(Mono.from(backend.save(bucketName, pushbackInputStream, storagePolicy)));
+    }
+
+    @Override
+    public BucketName getDefaultBucketName() {
+        return defaultBucket;
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, BlobId blobId) {
+        return Mono.from(backend.delete(bucketName, blobId))
+            .then(Mono.just(bucketName)
+                .filter(defaultBucket::equals)
+                .flatMap(ignored -> Mono.from(cache.remove(blobId)))
+                .then());
+    }
+
+    @Override
+    public Publisher<Void> deleteBucket(BucketName bucketName) {
+        return Mono.from(backend.deleteBucket(bucketName));
+    }
+
+    private Optional<byte[]> fullyReadSmallStream(PushbackInputStream pushbackInputStream) throws IOException {
+        int sizeToRead = sizeThresholdInBytes + 1;
```

Comment:
idem ?

Reply:
Can we adopt the same style than here: #3295 (comment)

Reply:
Yes, i doing it